packages/flutter/lib/src/services/raw_keyboard_linux.dart

```diff
@@ -85,7 +84,15 @@ class RawKeyEventDataLinux extends RawKeyEventData {
     // plane.
     if (keyLabel != null &&
         !LogicalKeyboardKey.isControlCharacter(keyLabel)) {
-      final int keyId = LogicalKeyboardKey.unicodePlane | (codePoint & LogicalKeyboardKey.valueMask);
+      // Not covering length > 2 case since > 1 is already unlikely.
+      assert(charactersIgnoringModifiers.length <= 2);
+      int codeUnit = charactersIgnoringModifiers.codeUnitAt(0);
+      if (charactersIgnoringModifiers.length == 2) {
+        final int secondCode = charactersIgnoringModifiers.codeUnitAt(1);
+        codeUnit = (codeUnit << 16) | secondCode;
+      }
```

Comment:
I don't understand this code at all.

It appears this code is untested. We should definitely test whatever this code is doing.

Reply:
Greg suggested this code as a safeguard against any really rare/custom keyboard that would send a unicode character with more than one code unit. I can add a test and documentation to clarify this.

Reply:
I'm not sure what you mean by "unicode character" or "code unit" or how a keyboard could do this. Can you elaborate?

Reply:
By "unicode character" I meant the code point. By "code unit" I was referring to the UTF-16 representation of the code point. This check covers the case were charactersIgnoringModifiers is a high code point and so it would be a surrogate pair.

This check might be a bit paranoid since it's highly unlikely that a keyboard would do this, but wanted to cover it anyway given the possibility in the future. I can remove it if you think it just adds complexity and little value.

Reply:
let's use the terms "Unicode Scalar Value" and "UTF-16 words" for those two terms, to maximally avoid ambiguity.

Regarding the code itself, IMHO we should either support arbitrary length strings, or we should only allow one Unicode scalar value, and if the latter, we must document this and assert it throughout (ideally by changing the API to not accept an arbitrary string at all, e.g. take a Unicode scalar value). We should definitely not have an API that looks like it takes an arbitrary string but doesn't.

FWIW, I think the code you have here is equivalent to just `codeUnit = codePoint.runes.single`.

Reply:
By suggesting the addition of handling the extra scalar value in the macOS code, I was trying to handle the possible situation when a single keystroke returns multiple scalar values. The docs for macOS's [charactersIgnoringModifiers] aren't very clear about how many values could be stuffed into the field, but our id scheme doesn't allow for more than two scalars, since it's only 32 bits.  Since we could accept up to two, I thought we should support up to the limit we could accept.  We should assert throughout that it doesn't contain more than two, however (the reading code does assert already if it's more than two).  Unfortunately, Dart doesn't have fixed-length arrays, or I'd also use one of those to indicate that the member can't have more than two characters, but we should document it. Using a packed int could work, however.

I'd rather not limit us to one value, since there may be some common cases (I'm thinking of combining characters here) where there are more than one, depending on how they handle those.

I'm not sure how clear the Linux docs are about this same situation, but if it's supplying a string, then it's probably similar enough to macOS to use a similar strategy.

And yes, it's equivalent to the first rune of the string (although I try and avoid working with runes, since it's such Dart specific terminology).

<!-- Links -->
[charactersIgnoringModifiers]: https://developer.apple.com/documentation/appkit/nsevent/1524605-charactersignoringmodifiers?language=objc